modules/llama_cpp_plugin/src/infer_request.cpp

```diff
@@ -0,0 +1,119 @@
+                          const ov::element::Type& element_type,
+                          const ov::Shape& shape) {
+    if (!tensor || tensor->get_element_type() != element_type) {
+        tensor = ov::make_tensor(element_type, shape);
+    } else {
+        tensor->set_shape(shape);
+    }
+}
+
+LlamaCppSyncInferRequest::LlamaCppSyncInferRequest(const std::shared_ptr<const LlamaCppModel>& compiled_model)
+    : ov::ISyncInferRequest(compiled_model) {
+    OPENVINO_DEBUG << "llama_cpp_plugin: infer request ctor called\n";
+    m_compiled_model_ptr = compiled_model;
+    for (const auto& input : get_inputs()) {
+        allocate_tensor(input, [input](ov::SoPtr<ov::ITensor>& tensor) {
+            allocate_tensor_impl(tensor,
+                                 input.get_element_type(),
+                                 input.get_partial_shape().is_dynamic() ? ov::Shape{0} : input.get_shape());
+        });
+    }
+    for (const auto& output : get_outputs()) {
+        allocate_tensor(output, [output](ov::SoPtr<ov::ITensor>& tensor) {
+            allocate_tensor_impl(tensor,
+                                 output.get_element_type(),
+                                 output.get_partial_shape().is_dynamic() ? ov::Shape{0} : output.get_shape());
+        });
+    }
+}
+void LlamaCppSyncInferRequest::set_tensors_impl(const ov::Output<const ov::Node> port,
+                                                const std::vector<ov::SoPtr<ov::ITensor>>& tensors) {
+    OPENVINO_DEBUG << "llama_cpp_plugin: set_tensors_impl called\n";
+}
+
+void llama_batch_add_reimpl(struct llama_batch& batch,
+                            llama_token id,
+                            llama_pos pos,
+                            const std::vector<llama_seq_id>& seq_ids,
+                            bool logits) {
+    batch.token[batch.n_tokens] = id;
+    batch.pos[batch.n_tokens] = pos;
+    batch.n_seq_id[batch.n_tokens] = seq_ids.size();
+    for (size_t i = 0; i < seq_ids.size(); ++i) {
+        batch.seq_id[batch.n_tokens][i] = seq_ids[i];
+    }
+    batch.logits[batch.n_tokens] = logits;
+
+    batch.n_tokens++;
+}
+
+void LlamaCppSyncInferRequest::infer() {
+    auto input_ids_tensor_ptr = get_tensor(get_inputs()[0]);  // TODO (vshampor) correctly identify input_ids among
+                                                              // all inputs without hardcode
+    OPENVINO_ASSERT(input_ids_tensor_ptr->get_element_type() == ov::element::Type_t::i64);
+    OPENVINO_ASSERT(input_ids_tensor_ptr->get_shape().size() == 2);
+    size_t sequence_length = input_ids_tensor_ptr->get_shape()[1];
+
+    // llama_batch actually contains one sequence
+    llama_batch batch = llama_batch_init(sequence_length, /* embd = */ 0, /* n_seq_max = */ 1);
+    const int64_t* data_ptr = input_ids_tensor_ptr->data<int64_t>();
+
+    const int64_t* sequence_start_ptr = data_ptr /* + seq_idx */;
+
+    for (size_t tok_idx = 0; tok_idx < sequence_length; ++tok_idx) {
```

Comment:
am I right that llama.cpp processes prompt token by token?

Reply:
This exact loop just adds tokens to the "llama.cpp batch", where the "batch" are the tokens for a single text sequence to be processed. `llama_batch_add_reimpl` is just a carbon copy of the internal helper function in llama.cpp https://github.com/ggerganov/llama.cpp/blob/2c4fb69246834503db7b78bcbedcef506bbc60c4/common/common.cpp#L1328, filling this struct can probably be done better, but I didn't want to move too far from llama.cpp at that moment.